### PR TITLE
[Merged by Bors] - refactor: use the typeclass `SProd` to implement overloaded notation `· ×ˢ ·`

### PR DESCRIPTION
**Pull request by Komyyy**
Currently, the following notations are changed from `· ×ˢ ·` because Lean 4 can't deal with ambiguous notations.
| Definition | Notation |
| :---: | :---: |
| `Finset.product` | `· ×ᶠ ·` |
| `LowerSet.prod` | `· ×ˡˢ ·` |
| `UpperSet.prod` | `· ×ᵘˢ ·` |

And following notations were still `· ×ˢ ·` but they causes performance issues; even timeouts.
| Definition | Notation |
| :---: | :---: |
| `Set.prod` | `· ×ˢ ·` |
| `List.product` | `· ×ˢ ·` |
| `Multiset.product` | `· ×ˢ ·` |

This PR makes the new typeclass `SProd` to implement overloaded notation `· ×ˢ ·`, and revert changed notations.
Also, this PR changes the notation of `Filter.prod` from `· ×ᶠ ·` to `· ×ˢ ·`.

---

- [x] depends on: #4308

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by ChrisHughes24 — APPROVED** [review]
I'm very happy with these changes, but they're quite sweeping so I'll let somebody else have an opinion as well.

**Comment by kmill** [comment]
I'm happy with these changes too. I know I had some reservations on Zulip, but I reviewed all the changes in closer detail and they're all improvements, so I say we go ahead.

Want to put it back at `awaiting-review` @Komyyy ? Like Chris the changes are quite sweeping and I'd like someone else to look at it (I'm fine with Chris deciding to merge it).

**Review by kmill — APPROVED** [review]

**Comment by Komyyy** [comment]
@kmill I'm merging your branch and removing unnecessary additional type ascriptions.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#4308
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Combinatorics/SimpleGraph/Prod.lean [hunk 1]
```diff
@@ -233,8 +232,7 @@ instance boxProdFintypeNeighborSet (x : α × β)
 theorem boxProd_neighborFinset (x : α × β)
     [Fintype (G.neighborSet x.1)] [Fintype (H.neighborSet x.2)] [Fintype ((G □ H).neighborSet x)] :
     (G □ H).neighborFinset x =
-      -- porting note: was `×ˢ`
-      (G.neighborFinset x.1 ×ᶠ {x.2}).disjUnion ({x.1} ×ᶠ H.neighborFinset x.2)
+      (G.neighborFinset x.1 ×ˢ {x.2}).disjUnion ({x.1} ×ˢ H.neighborFinset x.2)
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Prod.lean:235]: Very nice! I'm glad that we now have a metaprogrammatic way of solving this problem.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4308~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by YaelDillies** [comment]
Note that you could also use this notation on `Filter`.

**Comment by Komyyy** [comment]
@YaelDillies I'm working for this.

**Comment by semorrison** [comment]
bors d+

**Comment by urkud** [comment]
Historical remark: I introduced a similar typeclass in leanprover-community/mathlib#11300, then @YaelDillies replaced it with overlapping notations in leanprover-community/mathlib#15717 because it worked better in Lean 3.

**Comment by kmill** [comment]
@urkud To add to the historical remark: one thing that is different now from then is that instances of this `SProd` class make use of a custom `fbinop%` elaborator that give us similar benefits to the Lean 3 overlapping notations despite being a class-based design. (I pinged you and Yaël earlier in this thread to show a line that couldn't work with a Lean 3 class, but could work with Lean 3 overlapping notation, and now still works with Lean 4 classes.)

**Pull request closed without merging**